File: src/static/js/tabs/evaluate.js
/* evaluate.js — the Evaluate tab.
   Side-by-side generation, the style scorecard, the memorization check and the
   regression set. Held-out loss lives in the Train tab's curve; this is
   everything you can only see by actually generating. */

Forge.tab('evaluate', (() => {
  let mode = 'compare';

  function render(root) {
    const boot = Forge.state.boot || {};
    root.innerHTML = `
      <h1 class="page">Evaluate</h1>
      <p class="page-sub">Did the training take? Train loss going down doesn't answer that.
        These four checks do — comparison against the base model, measurable style match,
        memorization risk, and drift between runs.</p>

      <div class="row" style="margin-bottom:16px">
        ${[['compare', 'Side by side'], ['style', 'Style scorecard'],
           ['memo', 'Memorization'], ['regression', 'Regression set']].map(([m, l]) =>
          `<button class="btn sm${m === mode ? ' primary' : ''}" data-act="mode" data-m="${m}">${l}</button>`).join('')}
      </div>
      <div id="evalBody"></div>`;

    Forge.acts(root, {
      mode: (el) => { mode = el.dataset.m; render(root); },
      compare: () => runCompare(),
      style: () => runStyle(),
      memo: () => runMemo(),
      runReg: () => runRegression(),
      saveReg: () => saveRegression(),
    });
    paint();
  }

  const chars = () => Forge.state.boot?.characters || [];
  const dsets = () => Forge.state.boot?.datasets || [];

  function paint() {
    const body = document.getElementById('evalBody');
    if (!body) return;
    if (!chars().length) {
      body.innerHTML = `<div class="empty"><div class="big">◎</div>
        <p><strong>Nothing to evaluate yet.</strong></p>
        <p>Train an adapter and register it as a character first.</p></div>`;
      return;
    }
    if (mode === 'compare') return paintCompare(body);
    if (mode === 'style') return paintStyle(body);
    if (mode === 'memo') return paintMemo(body);
    if (mode === 'regression') return paintRegression(body);
  }

  // ── COMPARE ────────────────────────────────────────────
  function paintCompare(body) {
    body.innerHTML = `
      <div class="card">
        <div class="card-head"><h3>One prompt, several adapters</h3></div>
        <p class="hint" style="margin-bottom:12px">
          Only adapters loaded into the running engine can answer — restart the
          engine to include one that isn't. This is what the old
          <code>*_pre_dpo_backup</code> snapshots should have been for.</p>
        ${Forge.field('Prompt', `<textarea id="cmpPrompt" rows="2"
          placeholder="whats your take on people using ai to write all their code"></textarea>`)}
        <h3 class="sub">Compare</h3>
        <div class="row" style="margin-bottom:10px">
          ${chars().map(c => `<label class="check"><input type="checkbox" class="cmpChar"
            value="${c.id}" checked> ${Forge.esc(c.name)}</label>`).join('')}
          <label class="check"><input type="checkbox" id="cmpBase" checked> base model</label>
          <label class="check"><input type="checkbox" id="cmpPersona" checked> use persona prompt</label>
        </div>
        <div class="row">
          ${Forge.field('Temperature', Forge.num('cmpTemp', 0.8, { min: 0, max: 2, step: 0.05 }))}
          ${Forge.field('Max tokens', Forge.num('cmpTok', 200, { min: 32, max: 2048, step: 32 }))}
          <button class="btn primary" data-act="compare" style="margin-top:6px">Generate</button>
        </div>
      </div>
      <div id="cmpOut"></div>`;
  }

  async function runCompare() {
    const prompt = Forge.val('cmpPrompt').trim();
    if (!prompt) return Forge.err('Type a prompt first.');
    const ids = Forge.$$('.cmpChar').filter(c => c.checked).map(c => c.value);
    document.getElementById('cmpOut').innerHTML = '<p class="hint">Generating…</p>';
    const r = await Forge.post('/api/eval/compare', {
      prompt, char_ids: ids,
      include_base: Forge.chk('cmpBase'),
      use_persona: Forge.chk('cmpPersona'),
      temperature: Forge.numv('cmpTemp', 0.8),
      max_tokens: Forge.numv('cmpTok', 200),
    });
    document.getElementById('cmpOut').innerHTML = `<div class="split-scroll">${
      r.results.map(x => `
        <div class="card">
          <div class="card-head"><h3>${Forge.esc(x.label)}</h3><div class="spacer"></div>
            ${x.adapter_loaded ? '' : '<span class="pill crit">not loaded</span>'}
            <span class="pill">${x.seconds}s</span></div>
          <div style="font-size:13.5px;white-space:pre-wrap;line-height:1.6">${
            Forge.esc(x.text || x.error || '(nothing)')}</div>
          ${x.style ? `<div class="row tight" style="margin-top:10px">
            <span class="pill">${Math.round(x.style.lowercase_ratio * 100)}% lowercase</span>
            <span class="pill">${x.style.avg_sentence_words.toFixed(1)} w/sentence</span>
            <span class="pill">${x.style.words} words</span></div>` : ''}
        </div>`).join('')}</div>`;
  }

  // ── STYLE ──────────────────────────────────────────────
  function paintStyle(body) {
    body.innerHTML = `
      <div class="card">
        <div class="card-head"><h3>Does it sound like the training data?</h3></div>
        <p class="hint" style="margin-bottom:12px">
          Generates against the regression prompts, then measures the output against
          the dataset's own measurable style. A persona specified as "lowercase, drops
          apostrophes, sparse punctuation" is directly checkable — this checks it.</p>
        <div class="grid c3" style="gap:0 12px">
          ${Forge.field('Character', Forge.sel('stChar', chars().map(c => [c.id, c.name]),
            Forge.state.char).replace('id="__x"', 'id="stChar"'))}
          ${Forge.field('Compare against dataset', Forge.sel('stData', dsets().map(d => d.name),
            Forge.state.dataset).replace('id="__x"', 'id="stData"'))}
          ${Forge.field('Prompts to use', Forge.num('stN', 6, { min: 2, max: 30 }))}
        </div>
        <button class="btn primary" data-act="style">Run scorecard</button>
      </div>
      <div id="stOut"></div>`;
  }

  async function runStyle() {
    document.getElementById('stOut').innerHTML = '<p class="hint">Generating and measuring…</p>';
    const r = await Forge.post('/api/eval/style', {
      char_id: Forge.val('stChar'), dataset: Forge.val('stData'), n: Forge.numv('stN', 6),
    });
    const verdictPill = v => `<span class="pill ${v === 'close' ? 'ok' : v === 'drifting' ? 'warn' : 'crit'}">${v}</span>`;
    document.getElementById('stOut').innerHTML = `
      <div class="card">
        <div class="card-head"><h3>${Forge.esc(r.character)} vs ${Forge.esc(r.dataset)}</h3>
          <div class="spacer"></div>
          <span class="pill ${r.overall_distance < 0.25 ? 'ok' : r.overall_distance < 0.45 ? 'warn' : 'crit'}">
            overall distance ${r.overall_distance}</span></div>
        <div class="tw"><table><thead><tr><th>Dimension</th><th class="num">Dataset</th>
          <th class="num">Model</th><th class="num">Gap</th><th></th></tr></thead><tbody>
          ${r.dimensions.map(d => `<tr><td>${Forge.esc(d.label)}</td>
            <td class="num">${d.dataset}</td><td class="num">${d.model}</td>
            <td class="num">${d.gap}</td><td>${verdictPill(d.verdict)}</td></tr>`).join('')}
        </tbody></table></div>
      </div>
      <div class="card">
        <div class="card-head"><h3>What it actually said</h3></div>
        ${r.samples.map(s => `<div style="margin-bottom:14px">
          <div class="hint" style="margin:0 0 3px">${Forge.esc(s.prompt)}</div>
          <div style="font-size:13px;white-space:pre-wrap">${Forge.esc(s.text || s.error || '')}</div>
          ${s.distance != null ? `<span class="pill ${s.distance < 0.3 ? 'ok' : 'warn'}"
            style="margin-top:4px;display:inline-block">distance ${s.distance}</span>` : ''}
        </div>`).join('')}
      </div>`;
  }

  // ── MEMORIZATION ───────────────────────────────────────
  function paintMemo(body) {
    body.innerHTML = `
      <div class="card">
        <div class="card-head"><h3>Is it reciting its training data?</h3></div>
        <p class="hint" style="margin-bottom:12px">
          Answers prompts it was trained on, then finds the closest training row to
          each answer. Some overlap is expected — most of them near-verbatim means
          the adapter memorized rather than generalized, which is the real risk at
          a few hundred examples.</p>
        <div class="grid c3" style="gap:0 12px">
          ${Forge.field('Character', Forge.sel('mmChar', chars().map(c => [c.id, c.name]),
            Forge.state.char).replace('id="__x"', 'id="mmChar"'))}
          ${Forge.field('Dataset', Forge.sel('mmData', dsets().map(d => d.name),
            Forge.state.dataset).replace('id="__x"', 'id="mmData"'))}
          ${Forge.field('Similarity threshold', Forge.num('mmThr', 0.82, { min: 0.5, max: 0.99, step: 0.01 }))}
        </div>
        ${Forge.field('Prompts to check', Forge.num('mmN', 12, { min: 3, max: 40 }))}
        <button class="btn primary" data-act="memo">Run check</button>
      </div>
      <div id="mmOut"></div>`;
  }

  async function runMemo() {
    document.getElementById('mmOut').innerHTML = '<p class="hint">Generating and comparing…</p>';
    const r = await Forge.post('/api/eval/memorization', {
      char_id: Forge.val('mmChar'), dataset: Forge.val('mmData'),
      threshold: Forge.numv('mmThr', 0.82), n: Forge.numv('mmN', 12),
    });
    const kind = r.memorized_pct > 50 ? 'crit' : r.memorized_pct > 20 ? 'warn' : 'ok';
    document.getElementById('mmOut').innerHTML = `
      <div class="grid c4" style="margin-bottom:14px">
        <div class="stat"><div class="k">Checked</div><div class="v">${r.checked}</div></div>
        <div class="stat"><div class="k">Memorized</div>
          <div class="v" style="color:var(--${kind})">${r.memorized}</div>
          <div class="n">${r.memorized_pct}% of prompts</div></div>
        <div class="stat"><div class="k">Mean similarity</div><div class="v">${r.mean_similarity}</div></div>
        <div class="stat"><div class="k">Verdict</div><div class="v sm" style="color:var(--${kind})">${
          r.memorized_pct > 50 ? 'overfit' : r.memorized_pct > 20 ? 'watch it' : 'healthy'}</div></div>
      </div>
      ${r.findings.map(f => `
        <div class="card tight">
          <div class="card-head">
            <span class="pill ${f.memorized ? 'crit' : 'ok'}">${f.similarity}</span>
            <span class="hint" style="margin:0">${Forge.esc(f.prompt)}</span></div>
          <div class="grid c2">
            <div><div class="hint">generated</div>
              <div style="font-size:13px;white-space:pre-wrap">${Forge.esc(f.generated)}</div></div>
            <div><div class="hint">closest training row</div>
              <div style="font-size:13px;white-space:pre-wrap;color:var(--dim)">${
                Forge.esc(f.closest_training_row)}</div></div>
          </div>
        </div>`).join('')}`;
  }

  // ── REGRESSION ─────────────────────────────────────────
  async function paintRegression(body) {
    const p = await Forge.get('/api/eval/regression/prompts');
    body.innerHTML = `
      <div class="card">
        <div class="card-head"><h3>Pinned prompts</h3><div class="spacer"></div>
          <button class="btn sm" data-act="saveReg">Save list</button></div>
        <p class="hint" style="margin-bottom:10px">Re-run these after every training pass.
          Each run is diffed against the previous one, so you can see what a change
          actually changed.</p>
        <textarea id="regPrompts" class="mono" rows="9">${Forge.esc(p.prompts.join('\n'))}</textarea>
      </div>
      <div class="card">
        <div class="card-head"><h3>Run</h3><div class="spacer"></div>
          ${Forge.sel('regChar', chars().map(c => [c.id, c.name]), Forge.state.char)
            .replace('id="__x"', 'id="regChar" style="max-width:180px"')}
          <button class="btn primary sm" data-act="runReg">Run regression</button></div>
        <div id="regOut"></div>
      </div>`;
    loadRegHistory();
  }

  async function saveRegression() {
    const prompts = Forge.val('regPrompts').split('\n').map(s => s.trim()).filter(Boolean);
    await Forge.post('/api/eval/regression/prompts', { prompts });
    Forge.ok(`${prompts.length} prompts saved`);
  }

  async function runRegression() {
    document.getElementById('regOut').innerHTML = '<p class="hint">Running every prompt…</p>';
    const r = await Forge.post('/api/eval/regression/run', { char_id: Forge.val('regChar') });
    paintRegRun(r);
  }

  async function loadRegHistory() {
    const cid = Forge.val('regChar') || Forge.state.char;
    if (!cid) return;
    const runs = await Forge.get(`/api/eval/regression/${cid}`).catch(() => []);
    if (runs.length) paintRegRun(runs[runs.length - 1], runs.length);
  }

  function paintRegRun(r, count) {
    const changed = r.results.filter(x => x.changed).length;
    document.getElementById('regOut').innerHTML = `
      <div class="row" style="margin:10px 0">
        <span class="pill">${Forge.esc(r.label)}</span>
        <span class="pill ${changed ? 'warn' : 'ok'}">${changed} of ${r.results.length} changed</span>
        ${count ? `<span class="pill">run ${count}</span>` : ''}
      </div>
      ${r.results.map(x => `
        <div style="border-top:1px solid var(--line-2);padding:10px 0">
          <div class="hint" style="margin:0 0 4px">${Forge.esc(x.prompt)}</div>
          <div style="font-size:13px;white-space:pre-wrap">${Forge.esc(x.text || x.error || '')}</div>
          ${x.previous ? `<details style="margin-top:6px">
            <summary class="hint" style="cursor:pointer">previous${
              x.drift != null ? ` · drift ${x.drift}` : ''}</summary>
            <div style="font-size:12.5px;color:var(--dim);white-space:pre-wrap;margin-top:4px">${
              Forge.esc(x.previous)}</div></details>` : ''}
        </div>`).join('')}`;
  }

  return { render, enter() { paint(); } };
})());

File: src/core/evaluate.py
"""
core/evaluate.py — does the adapter actually work?

Four things the project had no answer for before:

  compare       one prompt, several adapters (and the bare base model) side by
                side, so "did DPO help or hurt" is answerable
  style_report  measure the outputs against the persona's own measurable style
                traits — lowercase ratio, apostrophe drops, sentence length
  memorization  flag generations that are near-verbatim training rows, the real
                risk when training 3 epochs on a few hundred examples
  regression    a pinned prompt set, re-run after each training pass and diffed
                against the previous run

Held-out loss is not here: it comes from the training run itself, via the
eval_split_pct config and the loss_history series in core/pipeline.py.
"""

import os
import json
import time
import difflib

import config
import dataset as ds
import store
from engine import MANAGER

REGRESSION_FILE = os.path.join(config.EVAL_DIR, "regression_prompts.json")
RUNS_DIR = os.path.join(config.EVAL_DIR, "runs")

DEFAULT_REGRESSION_PROMPTS = [
    "hey what are you up to",
    "whats your take on people using ai to write all their code",
    "i think we should just rewrite the whole thing from scratch",
    "explain how lora fine tuning works",
    "my build keeps failing and i have no idea why",
    "tell me something that happened to you recently",
    "whats the best thing you've built",
    "do you think local models will ever beat the big hosted ones",
]


# ─────────────────────────────────────────────────────────────
#  GENERATION HELPERS
# ─────────────────────────────────────────────────────────────

def _gen(prompt, char_id, persona="", temperature=0.8, max_tokens=200,
         enable_thinking=False):
    """One generation through the engine with a specific adapter active.
    char_id=None means the bare base model — that's the comparison baseline."""
    messages = []
    if persona:
        messages.append({"role": "system", "content": persona})
    messages.append({"role": "user", "content": prompt})
    return MANAGER.complete(messages, char_id=char_id, temperature=temperature,
                            max_tokens=max_tokens, enable_thinking=enable_thinking)


def compare(prompt, char_ids, temperature=0.8, max_tokens=200,
            include_base=True, use_persona=True):
    """Same prompt through several adapters. Returns one entry per variant.

    An adapter that isn't loaded into the running engine is reported as such
    rather than silently falling through to base-model output — that
    distinction is the entire point of this view.
    """
    if not MANAGER.is_ready():
        raise RuntimeError("Engine not running — start it from the Engine tab first.")

    variants = []
    if include_base:
        variants.append({"label": "base model", "char_id": None, "persona": ""})
    for cid in char_ids or []:
        char = store.get_character(cid)
        if not char:
            continue
        variants.append({
            "label": char.get("name", cid),
            "char_id": cid,
            "persona": char.get("persona", "") if use_persona else "",
        })

    loaded = {a["name"] for a in getattr(MANAGER, "adapters", [])}
    results = []
    for v in variants:
        entry = {"label": v["label"], "char_id": v["char_id"],
                 "adapter_loaded": v["char_id"] is None or v["char_id"] in loaded}
        t0 = time.time()
        try:
            if v["char_id"] and not entry["adapter_loaded"]:
                entry["text"] = ""
                entry["error"] = ("This character's adapter is not loaded in the "
                                  "running engine — restart the engine to include it.")
            else:
                entry["text"] = _gen(prompt, v["char_id"], v["persona"],
                                     temperature, max_tokens)
        except Exception as e:
            entry["text"] = ""
            entry["error"] = str(e)
        entry["seconds"] = round(time.time() - t0, 2)
        entry["style"] = ds.style_profile_text(entry.get("text", ""))
        results.append(entry)

    return {"prompt": prompt, "temperature": temperature, "results": results}


# ─────────────────────────────────────────────────────────────
#  STYLE SCORECARD
# ─────────────────────────────────────────────────────────────

STYLE_LABELS = {
    "lowercase_ratio":    "Lowercase ratio",
    "avg_word_len":       "Avg word length",
    "avg_sentence_words": "Words per sentence",
    "punct_density":      "Punctuation density",
    "apostrophe_drop":    "Apostrophe drop rate",
    "question_ratio":     "Questions per sentence",
    "exclaim_ratio":      "Exclamations per sentence",
}


def style_report(char_id, dataset_name, prompts=None, temperature=0.8,
                 max_tokens=200, n=None):
    """Generate against a set of prompts, then compare the output's measurable
    style against the training data's own style. A trained adapter should land
    close to its dataset; a big gap means the voice didn't transfer."""
    if not MANAGER.is_ready():
        raise RuntimeError("Engine not running — start it from the Engine tab first.")

    char = store.get_character(char_id)
    if not char:
        raise KeyError(f"No character {char_id}")

    reference = ds.dataset_style_profile(dataset_name)
    if not reference:
        raise ValueError(f"Dataset '{dataset_name}' is empty — nothing to compare against.")

    prompts = prompts or load_regression_prompts()
    if n:
        prompts = prompts[:int(n)]

    samples = []
    for p in prompts:
        try:
            text = _gen(p, char_id, char.get("persona", ""), temperature, max_tokens)
        except Exception as e:
            samples.append({"prompt": p, "text": "", "error": str(e)})
            continue
        samples.append({"prompt": p, "text": text,
                        "distance": round(ds.style_distance(text, reference), 3)})

    blob = "\n".join(s.get("text", "") for s in samples if s.get("text"))
    observed = ds.style_profile_text(blob) or {}

    dims = []
    for key, label in STYLE_LABELS.items():
        ref_v = reference.get(key)
        obs_v = observed.get(key)
        if ref_v is None or obs_v is None:
            continue
        scale = ds._STYLE_SCALE.get(key) or max(abs(ref_v), 1e-6)
        gap = min(1.0, abs(obs_v - ref_v) / scale)
        dims.append({
            "key": key, "label": label,
            "dataset": round(ref_v, 4), "model": round(obs_v, 4),
            "gap": round(gap, 3),
            "verdict": "close" if gap < 0.2 else ("drifting" if gap < 0.45 else "off"),
        })

    scored = [s["distance"] for s in samples if "distance" in s]
    return {
        "character": char.get("name", char_id),
        "char_id": char_id,
        "dataset": dataset_name,
        "overall_distance": round(sum(scored) / len(scored), 3) if scored else None,
        "dimensions": dims,
        "samples": samples,
    }


# ─────────────────────────────────────────────────────────────
#  MEMORIZATION
# ─────────────────────────────────────────────────────────────

def memorization_check(char_id, dataset_name, prompts=None, threshold=0.82,
                       temperature=0.8, max_tokens=200, n=12):
    """Is the adapter reciting its training data rather than generalising?

    Generates fresh answers, then finds the closest training output to each.
    Anything above the threshold is effectively memorised — expected to some
    degree, alarming if it's most of them.
    """
    if not MANAGER.is_ready():
        raise RuntimeError("Engine not running — start it from the Engine tab first.")

    char = store.get_character(char_id)
    if not char:
        raise KeyError(f"No character {char_id}")

    rows, _ = ds.read_rows(dataset_name)
    train_outputs = [r["output"].strip() for r in rows if r["output"].strip()]
    if not train_outputs:
        raise ValueError(f"Dataset '{dataset_name}' has no outputs to compare against.")

    # Prefer prompts the model was actually trained on — that's where
    # memorisation shows up, not on unseen prompts.
    prompts = prompts or [r["instruction"] for r in rows if r["instruction"].strip()][:int(n)]

    findings = []
    for p in prompts[:int(n)]:
        try:
            text = _gen(p, char_id, char.get("persona", ""), temperature, max_tokens).strip()
        except Exception as e:
            findings.append({"prompt": p, "error": str(e)})
            continue
        best, best_score = "", 0.0
        for t in train_outputs:
            score = difflib.SequenceMatcher(None, text.lower(), t.lower()).ratio()
            if score > best_score:
                best, best_score = t, score
        findings.append({
            "prompt": p, "generated": text,
            "closest_training_row": best,
            "similarity": round(best_score, 3),
            "memorized": best_score >= threshold,
        })

    scored = [f for f in findings if "similarity" in f]
    n_mem = sum(1 for f in scored if f["memorized"])
    return {
        "character": char.get("name", char_id),
        "dataset": dataset_name,
        "threshold": threshold,
        "checked": len(scored),
        "memorized": n_mem,
        "memorized_pct": round(100 * n_mem / len(scored), 1) if scored else 0,
        "mean_similarity": round(sum(f["similarity"] for f in scored) / len(scored), 3) if scored else 0,
        "findings": sorted(scored, key=lambda f: -f["similarity"]),
    }


# ─────────────────────────────────────────────────────────────
#  REGRESSION SET
# ─────────────────────────────────────────────────────────────

def load_regression_prompts():
    if os.path.exists(REGRESSION_FILE):
        try:
            with open(REGRESSION_FILE, encoding="utf-8") as f:
                data = json.load(f)
            if isinstance(data, list) and data:
                return data
        except (OSError, json.JSONDecodeError):
            pass
    return list(DEFAULT_REGRESSION_PROMPTS)


def save_regression_prompts(prompts):
    os.makedirs(config.EVAL_DIR, exist_ok=True)
    clean = [str(p).strip() for p in prompts if str(p).strip()]
    tmp = REGRESSION_FILE + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(clean, f, ensure_ascii=False, indent=2)
    os.replace(tmp, REGRESSION_FILE)
    return clean


def _runs_path(char_id):
    os.makedirs(RUNS_DIR, exist_ok=True)
    return os.path.join(RUNS_DIR, f"{char_id}.json")


def list_runs(char_id):
    p = _runs_path(char_id)
    if not os.path.exists(p):
        return []
    try:
        with open(p, encoding="utf-8") as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError):
        return []


def run_regression(char_id, temperature=0.8, max_tokens=200, label=None):
    """Run every pinned prompt and store the result, diffed against last time."""
    if not MANAGER.is_ready():
        raise RuntimeError("Engine not running — start it from the Engine tab first.")

    char = store.get_character(char_id)
    if not char:
        raise KeyError(f"No character {char_id}")

    prompts = load_regression_prompts()
    previous = list_runs(char_id)
    prev_map = {}
    if previous:
        prev_map = {r["prompt"]: r["text"] for r in previous[-1].get("results", [])}

    results = []
    for p in prompts:
        try:
            text = _gen(p, char_id, char.get("persona", ""), temperature, max_tokens)
        except Exception as e:
            results.append({"prompt": p, "text": "", "error": str(e)})
            continue
        prev = prev_map.get(p)
        results.append({
            "prompt": p,
            "text": text,
            "previous": prev,
            "changed": (prev is not None and prev.strip() != text.strip()),
            "drift": (round(1 - difflib.SequenceMatcher(None, prev.lower(), text.lower()).ratio(), 3)
                      if prev else None),
        })

    run = {
        "ts": time.time(),
        "label": label or time.strftime("%Y-%m-%d %H:%M"),
        "char_id": char_id,
        "character": char.get("name", char_id),
        "adapter": char.get("adapter_gguf", ""),
        "temperature": temperature,
        "results": results,
    }
    runs = previous + [run]
    runs = runs[-25:]          # keep the last 25 runs, not every one ever
    p = _runs_path(char_id)
    tmp = p + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(runs, f, ensure_ascii=False, indent=1)
    os.replace(tmp, p)
    return run


# ─────────────────────────────────────────────────────────────
#  TRAINING-RUN HISTORY
# ─────────────────────────────────────────────────────────────

def save_training_run(lora_name, status_snapshot):
    """Archive a finished run's loss history so the Evaluate tab can show
    curves for past runs, not only the one still in memory."""
    os.makedirs(RUNS_DIR, exist_ok=True)
    p = os.path.join(RUNS_DIR, f"_train_{os.path.basename(str(lora_name))}.json")
    history = []
    if os.path.exists(p):
        try:
            with open(p, encoding="utf-8") as f:
                history = json.load(f)
        except (OSError, json.JSONDecodeError):
            history = []
    history.append({
        "ts": time.time(),
        "lora_name": lora_name,
        "final_loss": status_snapshot.get("loss"),
        "final_eval_loss": status_snapshot.get("eval_loss"),
        "steps": status_snapshot.get("total_steps"),
        "epochs": status_snapshot.get("epoch"),
        "loss_history": status_snapshot.get("loss_history") or [],
    })
    history = history[-15:]
    tmp = p + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(history, f, ensure_ascii=False)
    os.replace(tmp, p)
    return len(history)


def training_runs(lora_name):
    p = os.path.join(RUNS_DIR, f"_train_{os.path.basename(str(lora_name))}.json")
    if not os.path.exists(p):
        return []
    try:
        with open(p, encoding="utf-8") as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError):
        return []
